src/libOpenImageIO/imagebufalgo_opencv.cpp

```diff
@@ -136,7 +137,182 @@ IplImage *
 ImageBufAlgo::to_IplImage (const ImageBuf &src)
 {
 #ifdef USE_OPENCV
-    return NULL;
+    ImageSpec spec = src.spec();
+    ImageBuf tmp = src;
+
+    // If the image has 4+ channels, then reduce to 4 channels, and use BGRA
+    // order (OpenCV ordering).
+    if (spec.nchannels >=3) {
+        int channelorder[spec.nchannels];
+
+        // Set the channel order
+        for (int i = 0; i < spec.nchannels; ++i) {
+            channelorder[i] = i;
+        }
+
+        channelorder[0] = 2;  // B
+        channelorder[1] = 1;  // G
+        channelorder[2] = 0;  // R
+
+        if (!ImageBufAlgo::channels(tmp, src, spec.nchannels, &channelorder[0])) {
+            DASSERT (0 && "Could not swap channels.");
+            return NULL;
+        }
+    }
+
+    // Create an IplImage to write to.
+    CvSize size = cvSize(spec.width, spec.height);
+    IplImage *dst;
+
+    // Get the pixel data from the ImageBuf and send it to the IplImage.
+    switch (spec.format.basetype) {
+        case TypeDesc::UINT8: {
+            dst = cvCreateImageHeader(size, IPL_DEPTH_8U, spec.nchannels);
+
+            if (!dst) {
+                DASSERT (0 && "Could not create dst IplImage.");
+                return NULL;
+            }
+
+            std::vector<unsigned char> data;
+            data.resize(spec.width * spec.height * spec.depth * spec.nchannels);
+
+            if (!tmp.get_pixels(ROI::All(), TypeDesc::UINT8, &data[0])) {
```

Comment:
I think that in many cases, there is an extra copy happening here as well. If spec.format.basetype == UINT8, then the ImageBuf's data is already unsigned char, so why are we calling get_pixels to copy it? That's only necessary if the data are non-local somehow, e.g. the image is backed by ImageCache, which can only happen for <=2 channel images since for >=3 channels, we made a copy (which will always be 'local'). So it seems to me that we can almost always eliminate this copy by directly testing the important case:

```
    if (spec.format.basetype == TypeDesc::UINT8 && tmp.storage == ImageBuf::IMAGECACHE) {
        // Type is right and the IB is not backed by an ImageCache, so
        // directly read from the local data of the IB.
        cvSetData (dst, tmp.localpixels(), dst->widthStep);
    else {
        // Either we are backed by an ImageCache, or we somehow still need a data
        // format conversion, so make a local copy with get_pixels.
        std::vector<unsigned char> data;
        data.resize (spec.width * spec.height * spec.depth * spec.nchannels);
        tmp.get_pixels (ROI::All(), TypeDesc::INT8, &data[0]);
        cvSetData(dst, &data[0], dst->widthStep);
    }
```

Voila, now we have no unnecessary pixel copy!